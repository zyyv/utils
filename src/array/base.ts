import { isArray } from '../is'
import type { Arrayable, Nullable } from '../types'

/**
 * Convert `Arrayable<T>` to `Array<T>`
 * 转换为一个数组
 *
 * @category Array
 */
export function toArray<T>(array?: Nullable<Arrayable<T>>): Array<T> {
  array = array || []
  if (isArray(array))
    return array
  return [array]
}

/**
 * Unique an Array
 * 转换为一个独一无二的数组
 *
 * @category Array
 */
export function uniq<T>(array: readonly T[]): T[] {
  return Array.from(new Set(array))
}

/**
 * Get nth item of Array. Negative for backward
 * 获取数组的第n个元素，负数表示倒数
 *
 * @category Array
 */
export function at(array: readonly [], index: number): undefined
export function at<T>(array: readonly T[], index: number): T
export function at<T>(array: readonly T[] | [], index: number): T | undefined {
  const len = array.length
  if (!len)
    return undefined

  if (index < 0)
    index += len

  return array[index]
}

/**
 * Get last item
 * 获取最后一个元素
 * @category Array
 */
export function last(array: readonly []): undefined
export function last<T>(array: readonly T[]): T
export function last<T>(array: readonly T[]): T | undefined {
  return at(array, -1)
}

/**
 * Remove an item from Array
 * 从数组中移除一个元素
 * @category Array
 */
export function remove<T>(array: T[], value: T): boolean {
  if (!array)
    return false
  const index = array.indexOf(value)
  if (index >= 0) {
    array.splice(index, 1)
    return true
  }
  return false
}

/**
 * Genrate a range array of numbers. The `stop` is exclusive.
 * 生成一个范围数组，包含起始值，但不包含结束值
 * @category Array
 */
export function range(stop: number): number[]
export function range(start: number, stop: number, step?: number): number[]
export function range(...args: any): number[] {
  let start: number, stop: number, step: number

  if (args.length === 1) {
    start = 0
    step = 1;
    ([stop] = args)
  }
  else {
    ([start, stop, step = 1] = args)
  }

  const arr: number[] = []
  let current = start
  while (current < stop) {
    arr.push(current)
    current += step || 1
  }

  return arr
}

/**
 * Move element in an Array
 * 移动数组中的元素
 *
 * @category Array
 */
export function move<T>(arr: T[], from: number, to: number): T[] {
  arr.splice(to, 0, arr.splice(from, 1)[0])
  return arr
}

/**
 * averaging
 * 求平均数
 * @param nums number
 * @returns number
 * @category Array
 */
export function average(...arr: number[]): number {
  return arr.reduce((c, n) => c + n, 0) / arr.length
}

/**
 * Randomly sorts the elements of an array using the Fisher-Yates algorithm.
 * 随机排序数组
 *
 * @category Array
 */
export function shuffle<T extends any[]>(arr: T): T {
  let m = arr.length
  while (m) {
    const i = Math.floor(Math.random() * m--) as number
    [arr[m], arr[i]] = [arr[i], arr[m]]
  }
  return arr
}

/**
 * Split array into many arrays
 * 将数组均分多个数组
 *
 * @category Array
 */
export function chunk<T>(arr: T[], size: number): T[][] {
  const result: T[][] = []
  for (let i = 0; i < arr.length; i += size)
    result.push(arr.slice(i, i + size))

  return result
}

/**
 * Array Map
 *
 * @category Array
 */
export default {
  toArray,
  uniq,
  at,
  last,
  remove,
  range,
  move,
  average,
  shuffle,
  chunk,
}
