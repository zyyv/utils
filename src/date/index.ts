import base from './base'

export * from './base'

/**
 * Date utils
 * @category Date
 */
export const UDate = {
  ...base,
}
