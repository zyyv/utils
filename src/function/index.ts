export * from './events'
export * from './pubsub'

/**
 * noop 空函数
 *
 * @category Function
 */
export const noop = () => {}

/**
 * no 永远返回false
 *
 * @category Function
 */
export const NO = () => false

/**
 * A function that is called only once
 * 只调用一次的函数
 * @category Function
 * @param fn (...args: any) => void
 * @returns Function
 */
export function once(fn: (...args: any) => void): Function {
  // 只调用一次的锁
  let called = false
  return function(...args: any) {
    if (!called) {
      called = true
      return fn(...args)
    }
  }
}
