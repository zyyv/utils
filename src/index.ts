export * from './array'
export * from './object'
export * from './is'
export * from './function'
export * from './types'
export * from './date'
export * from './shared'
export * from './number'
export * from './string'
export * from './promise'
