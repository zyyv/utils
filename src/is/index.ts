import env from './env'
import is from './is'

export * from './is'
export * from './env'

export const UIs = {
  ...env,
  ...is,
}
