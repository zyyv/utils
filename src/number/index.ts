import base from './base'

export * from './base'

export const UNumber = {
  ...base,
}
