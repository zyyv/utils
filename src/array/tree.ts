import { isValue } from '../is'
import type { Nullable } from '../types'

/**
 * Generate tree structure according by prop
 * 根据指定的属性生成树状结构
 *
 * eg: [{id: 1, pid: 0}, {id: 2, pid: 1}, {id: 3, pid: 1}]
 *
 * => [{id: 1, pid: 0, children: [{id: 2, pid: 1}, {id: 3, pid: 1}]}]
 * @param arr Array
 * @param id Nullable<number | string>
 * @param prop Nullable<number | string>
 * @category Array
 */
export function nest(arr: any[]): any[]
export function nest(arr: any[], id: Nullable<number | string>): any[]
export function nest(arr: any[], id: Nullable<number | string>, prop: Nullable<number | string>): any[]
export function nest(...args: any): any[] {
  const [arr] = args
  let [, id, prop] = args
  id = id ?? 0
  prop = prop ?? 'pid'
  return arr
    .filter((i: any) => isValue(i[prop]) && i[prop] === id)
    .map((i: any) => ({ ...i, children: nest(arr, i.id) }))
}

export default {
  nest,
}
