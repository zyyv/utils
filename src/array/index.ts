import base from './base'
import tree from './tree'

export * from './base'
export * from './tree'

/**
 * Array utils
 *
 * @category Array
 */
export const UArray = {
  ...base,
  ...tree,
}
