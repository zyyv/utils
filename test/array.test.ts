import { describe, expect, it } from 'vitest'
import { average, chunk, nest, range, shuffle, toArray } from '../src'
import { nest as nestData } from './targets/array'

describe('test Array utils', () => {
  it('toArray', () => {
    expect(toArray([0])).toEqual([0])
  })

  it('range', () => {
    expect(range(0)).toEqual([])
    expect(range(2)).toEqual([0, 1])
    expect(range(2, 5)).toEqual([2, 3, 4])
    expect(range(2, 10, 2)).toEqual([2, 4, 6, 8])
  })

  it('nest', () => {
    nestData.forEach(({ arr, id, pid }) => {
      expect(nest(arr, id, pid)).toMatchSnapshot()
    })
  })

  it('average', () => {
    expect(average(1, 2, 3, 4)).toBe(2.5)
    expect(average(...[2, 3, 4])).toBe(3)
  })

  it('shuffle', () => {
    expect(shuffle([1, 2, 3])).toHaveLength(3)
    expect(shuffle([{ a: 1 }, { b: 2 }, { c: 3 }])).toHaveLength(3)
  })

  it('chunk', () => {
    expect(chunk(range(10), 4)).toMatchSnapshot()
  })
})
