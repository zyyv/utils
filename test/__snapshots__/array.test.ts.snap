// Vitest Snapshot v1

exports[`test Array utils > chunk 1`] = `
[
  [
    0,
    1,
    2,
    3,
  ],
  [
    4,
    5,
    6,
    7,
  ],
  [
    8,
    9,
  ],
]
`;

exports[`test Array utils > nest 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "id": 2,
        "pid": 1,
      },
      {
        "children": [],
        "id": 3,
        "pid": 1,
      },
    ],
    "id": 1,
    "pid": 0,
  },
]
`;

exports[`test Array utils > nest 2`] = `
[
  {
    "children": [
      {
        "children": [],
        "id": 2,
        "pid": 1,
      },
      {
        "children": [],
        "id": 3,
        "pid": 1,
      },
    ],
    "id": 1,
    "pid": "1",
  },
]
`;

exports[`test Array utils > nest 3`] = `
[
  {
    "children": [],
    "id": 1,
    "parent": 0,
  },
]
`;

exports[`test Array utils > nest 4`] = `
[
  {
    "children": [],
    "id": 1,
    "parent": "1",
  },
]
`;
